### ISSUE TITLE
Allow overriding default SSL/TLS version support for load balancers

### ISSUE THREAD

**Issue by aldeed** [rancher/rancher#3097]
Thank you for the new haproxy config abilities, but unfortunately since your template sets `no-sslv3 no-tlsv10` on each bind, there does not seem to be any way to override that. Can you use a global `ssl-default-bind-options no-sslv3 no-tlsv10` instead, and allow us to override by adding for example `ssl-default-bind-options no-sslv3` in the "Global" section, which would replace your default line and therefore enable TLS 1.0? (Our end goal is just to enable 1.0 a little while longer while until all our clients support 1.2)

**Comment by rubenmch** [comment]
Are there any workarounds to enable tlsv10 or sslv3 in the current version of rancher?

Is this going to be added soon or until v1.0 is released?

We're currently connecting to the container and manually editing /etc/haproxy/haproxy.cfg every time we make some changes to the services or load balancer settings but this is error prone.

Can you recommend a way to automate editing the /etc/haproxy/haproxy.cfg file to remove no-tlsv10 or no-sslv3?

Is there any way to add some commands to be run when the load balancer container is restarted?

I tried the following unsuccessfully:

1. Using `crontab` (cron it's not running in the LB container)
2. Adding a service conf to `/etc/init` (it's not run either)
3. Editing `haproxy_start` and `haproy_reload` in `/etc/init.d/haproxy` (this works if I manually run `service haproxy reload` or `service haproxy start` but it seems rancher doesn't use that to reload or start haproxy. How does rancher reload or start haproxy?
4. Editing `/etc/init.d/agent-instance-startup` (the contents of this file are replaced each time the load balancer is restarted so the changes I made are lost)
5. Editing `/etc/inittab` (the contents of this file are replaced each time the load balancer is restarted so the changes I made are lost)

**Comment by sangeethah** [comment]
Tested with latest build from master:

Created  LB service with ssl enabled listeners:

We see the following being included in the global section in haproxy.conf:
``` ssl-default-bind-options no-sslv3 no-tlsv10 ```

Front end does not include ```no-sslv3 no-tlsv10``` option any more.

**Comment by sangeethah** [comment]
@aldeed , This should be available in the next rancher-server release.

**Comment by aldeed** [comment]
Excellent, thanks!

**Comment by aldeed** [comment]
@sangeethah We upgraded and this does not work. From what I can tell, `ssl-default-bind-options` was introduced in haproxy 1.5.7 and the LB is using haproxy 1.5.4. It seems like https://github.com/rancher/rancher/issues/3499 will fix this, but is there a way we can bump the haproxy version used on Rancher 0.56.1 LB?

**Comment by deniseschannon** [comment]
@aldeed This has not been released in Rancher yet. It will be in the next release, which will be coming out in the next week or two.

**Comment by aldeed** [comment]
OK gotcha thanks

**Comment by rubenmch** [comment]
I tested this in rancher 0.59.0 and it seems there's still no way to override the default `ssl-default-bind-options`.

When specifying `ssl-default-bind-options no-sslv3` in the global options section of the load balancer service rancher includes both the default options and the new options in haproxy.cfg:

```
ssl-default-bind-options no-sslv3 no-tlsv10
...
ssl-default-bind-options no-sslv3
```

This still doesn't enable TLS 1.0 since there aren't any options in haproxy to disable the no-tlsv10 option included by default.

The only option I've found is to use `force-tlsv10` but that forces all the clients to use TLS 1.0 even the ones that support TLS 1.1 or TLS 1.2.

Is there any other way in the current version of rancher version to support both TLS 1.0 and TLS 1.2 clients with a single load balancer?

**Comment by aldeed** [comment]
Rancher should look at any directives we add line by line and remove each corresponding line from the default defaults so that ours take effect. Or to put it another way, it should start with user-provided defaults and then add default defaults if not provided by the user.

**Comment by alena1108** [comment]
@rubenmch @aldeed reopening the bug.

**Comment by marcqualie** [comment]
Are there any plans or ETA for this? We currently need to support TLS 1.0 for some older clients using legacy windows software (out of our control) to connect to our APIs. We currently have rancher running all of our staging and QA systems but would love to have this feature when Rancher goes 1.0 for production use.

**Comment by sserrano44** [comment]
This affects every android under 4.4

**Comment by darrencrossley** [comment]
This caught us out too - to clarify the only way to enable TLSv1 is currently to shell-in to each LB and remove the lines from haconfig?

I'm on the latest Rancher - v1.0.1 and it's become quite apparent we have quite a few users on IE10 without admin privileges or TSL1.1 enabled. :-(

**Comment by rubenmch** [comment]
@darrencrossley You can also add the following to the global options of your load balancer in rancher, but then all the connections from all the clients will be made using TLS 1.0
```
ssl-default-bind-options force-tlsv10
```
Or you can create two separate load balancers in different hosts with different domains, one for TLS 1.0 clients and the other for the rest of the clients.

**Comment by weitzj** [comment]
There really needs to be a better fix for this.

- Android < 4.4 browsers do not handle TLS1.2, but need TLS1.0. (TLS 1.2. is still possible in your own apps though)
- iOS needs better TLS than TLS1.0

**Comment by andrewcarpenter** [comment]
Our cheesy "solution" here to customize the load balancer config was to add the following to the load-balancer container:

new file: `/var/lib/cattle/modify_haproxy_config.sh`:
```bash
set -e

modify_haproxy_config() {
  local config_path=$1/etc/haproxy/haproxy.cfg

  if [ -e $config_path ]; then
    sed -i \
      -e 's|ssl-default-bind-options no-sslv3 no-tlsv10|ssl-default-bind-options no-sslv3|' \
      -e "s|bind \*:443  ssl crt /etc/haproxy/certs/|&\\n        http-response set-header Strict-Transport-Security \"max-age=15552000; includeSubDomains; preload;\"|" \
      -e "s|bind \*:80|&\\n        redirect scheme https code 301 if \!\{ ssl_fc \}|" \
      $config_path
  fi
}
```

(note this both updates the TLS versions supported, adds HSTS, and redirects port 80 to 443)

and add the following to the end of the `download` function in `/var/lib/cattle/config.sh`:

```bash
    source $(dirname $0)/modify_haproxy_config.sh
    modify_haproxy_config "${content_root}/content/"
```

by punching it into the load balancer configuration update functionality, the changes get automatically applied when you add new hosts or re-deploy a service.

This is definitely hacky, but works for us while we wait for more flexibility in load balancer configuration.

**Comment by weitzj** [comment]
What we are using now is:

- Do TLS termination in front of haproxy. 
- Use haproxy only as an internal load balancer "lb-internal"
- This https://github.com/GoogleCloudPlatform/nginx-ssl-proxy image works for us now. Certs are put on the host via puppet.
-image does port 80 redirect 301 to 443 as well
- run image inside the rancher overlay network
- as a target host use the name of your internal lb, "lb-internal"

- still thinking to use hitch https://github.com/varnish/hitch with PROXY protocol support 

- this does well for HTTP proxying. TCP proxying with SNi us still unsolved

**Comment by stevenschobert** [comment]
Similar to what @darrencrossley said, here are the commands I'm running manually in the load balancer shell, while waiting on this to be addressed in a future release (in case it helps anyone else):

```sh
#optional, if you want to bkup before changing
cp /etc/haproxy/haproxy.cfg /etc/haproxy/haproxy.cfg.bkup 

# replace line in file
cat /etc/haproxy/haproxy.cfg | sed 's/ssl-default-bind-options no-sslv3 no-tlsv10/ssl-default-bind-options no-sslv3/' > /etc/haproxy/haproxy.cfg

# restart haproxy
/etc/init.d/haproxy restart
```
Definitely not a robust solution, but if you are running a small setup like me, its not much work to just shell and change it manually.

**Comment by outrunthewolf** [comment]
Is the source doe for this o/s? I'd be happy to look into a patch or p/r

**Comment by mishak87** [comment]
@outrunthewolf source code of container running haproxy is in [rancher/agent-instance](https://github.com/rancher/agent-instance) repository.

**Comment by mishak87** [comment]
[My solution](https://gist.github.com/mishak87/a236a53844ffcf50eaa2f21d72f192e0) is based on Andrews but without need for manually patching each container. Use on your own risk. Comments are welcome.

**Comment by boshido** [comment]
Was this problem already been fixed in 1.2.0-pre2 ?

**Comment by alena1108** [comment]
https://github.com/rancher/rancher/issues/2179

**Comment by sangeethah** [comment]
Tested with build from master and following micro service versions:
rancher/dns:v0.10.5
rancher/healthcheck:v0.1.0
rancher/net:v0.7.0
rancher/metadata:v0.6.3
rancher/plugin-manager:v0.2.1
rancher/scheduler:v0.1.2

When creating  lb service set the following value in "Custom haproxy.cfg" tab:
```
global
ssl-default-bind-options no-sslv3
```
haproxy config will now have only ```ssl-default-bind-options no-sslv3``` in global section:

<img width="695" alt="screen shot 2016-11-16 at 5 03 03 pm" src="https://cloud.githubusercontent.com/assets/4266958/20372315/0ad597a2-ac1f-11e6-8375-fe9d1fcb540f.png">

This specific issue of overriding existing config parameters was fixed and verified in https://github.com/rancher/rancher/issues/5745
 
@rubenmch @aldeed , Fix for this issue will be available in 1.2 release.
